sw/device/lib/crypto/api.h

```diff
@@ -0,0 +1,1378 @@
+ *
+ * @param drbg_state Pointer to the drbg working state
+ * @param additional_input Pointer to the additional input for drbg
+ * @param required_key_len Requested key length in bits
+ * @param N Pointer to un-blinded key struct with modulus (N)
+ * @param E Pointer to un-blinded key struct with public exponent (E)
+ * @param D Pointer to blinded key struct with private exponent (D)
+ * @return crypto_status_t The result of the rsa key generation
+ */
+crypto_status_t rsa_keygen_async(drbg_state_t *drbg_state,
+                                 crypto_uint8_buf_t additional_input,
+                                 size_t required_key_len,
+                                 crypto_unblinded_key_t *N,
+                                 crypto_unblinded_key_t *E,
+                                 crypto_blinded_key_t *D);
+
+/**
+ * Computes the digital signature on the input message data, async.
+ *
+ * @param N Pointer to un-blinded key struct with modulus (N)
+ * @param D Pointer to blinded key struct with private exponent (D)
+ * @param input_message Input message to be signed
+ * @param padding_mode Padding scheme to be used for the data
+ * @param hash_mode Hashing scheme to be used for the signature scheme
+ * @param signature Pointer to generated signature struct
+ * @return crypto_status_t The result of the rsa sign generation
+ */
+crypto_status_t rsa_sign_async(crypto_unblinded_key_t *N,
+                               crypto_blinded_key_t *D,
+                               const crypto_uint8_buf_t input_message,
+                               rsa_padding_t padding_mode, rsa_hash_t hash_mode,
+                               crypto_uint8_buf_t *signature);
+
+/**
+ * Verifies the authenticity of the input signature, asynchronously.
+ *
+ * The generated signature is compared against the input signature and
+ * PASS / FAIL is returned
+ *
+ * @param N Pointer to un-blinded key struct with modulus (N)
+ * @param E Pointer to un-blinded key struct with public exponent (E)
+ * @param input_message Input message to be signed for verification
+ * @param padding_mode Padding scheme to be used for the data
+ * @param hash_mode Hashing scheme to be used for the signature scheme
+ * @param signature Pointer to the signature to be verified
+ * @param verification_result Returns the result of signature verification
+ * (Pass/Fail)
+ * @return crypto_status_t The return status of the rsa verify operation
+ */
+crypto_status_t rsa_verify_async(crypto_unblinded_key_t *N,
+                                 crypto_unblinded_key_t *E,
+                                 const crypto_uint8_buf_t input_message,
+                                 rsa_padding_t padding_mode,
+                                 rsa_hash_t hash_mode,
+                                 crypto_uint8_buf_t signature,
+                                 verification_status_t *verification_result);
+
+/**
+ * Sets the domain parameters of a generic (Weierstrass) ECC curve.
```

Comment:
Presumably this mutates some global state, right? Can we document that?

Reply:
There's already an open comment on the document about this; I agree it should be either clearly documented or non-global. But I'm reluctant to make a call here before the discussion in the main doc is resolved...

Reply:
Ack